descriptor.md

```diff
@@ -41,20 +41,20 @@ The following fields contain the primary properties that constitute a Descriptor
 
 - **`annotations`** *string-string map*
 
-    This OPTIONAL property contains arbitrary metadata for this descriptor.
-    This OPTIONAL property MUST use the [annotation rules](annotations.md#rules).
+  This OPTIONAL property contains arbitrary metadata for this descriptor.
+  This OPTIONAL property MUST use the [annotation rules](annotations.md#rules).
 
-Descriptors pointing to [`application/vnd.oci.image.manifest.v1+json`](manifest.md) SHOULD include the extended field `platform`, see [Image Index Property Descriptions](image-index.md#image-index-property-descriptions) for details.
-
-### Reserved
+- **`data`** *string*
 
-The following field keys are reserved and MUST NOT be used by other specifications.
+  This OPTIONAL property contains an embedded representation of the referenced content.
+  Values MUST conform to the Base 64 encoding, as defined in [RFC 4648][rfc4648-s4].
+  The decoded data MUST be identical to the referenced content and SHOULD be verified against the [`digest`](#digests) and `size` fields by content consumers.
```

Comment:
If the provided data field _doesn't_ match the digest, should clients download the blob and try that, or just assume the blob is no good as if they had downloaded it?

(IMO it seems like it would be useful to have a canonical answer to what clients "SHOULD" do in this situation)

Reply:
I guess on the flip side, can/"SHOULD" a registry reject a manifest if the embedded data does not appropriately match the digest?  Could it also let the client then skip uploading that blob directly by ingesting this data field into the appropriate blob instead of requiring it be uploaded prior?

Reply:
> I guess on the flip side, can/"SHOULD" a registry reject a manifest if the embedded data does not appropriately match the digest? Could it also let the client then skip uploading that blob directly by ingesting this data field into the appropriate blob instead of requiring it be uploaded prior?

I'd lean towards some MAY language for that. For backwards compatibility reasons, you generally want to upload the blob in addition to embedding it, because older clients won't know about `data`. Similarly, registries that don't look at `data` won't validate anything, so IMO it's better that we assume the registry is dumb and hasn't done this kind of validation.

If a registry _does_ want to allow skipping the upload for embedded content, then it SHOULD validate that the contents match, but I would be really reluctant to recommend this until we've let this change bake for a while (years).

Reply:
> If the provided data field doesn't match the digest, should clients download the blob and try that, or just assume the blob is no good as if they had downloaded it?

Good question. This feels similar to the `urls` field.

> This OPTIONAL property specifies a list of URIs from which this object MAY be downloaded.

We don't prescribe any order in which content sources should be tried, just that the _object MAY be downloaded_ from the URI. If we want to define some fallback behavior, we should include `urls` too.

I would expect the order of operations to be something like:
1. Try `data`.
2. Try the contextual content store (registry, blobs/ directory, Merkle DAG in the sky).
3. Try the `urls` addresses.

In code I've written, I only fall back to `urls` if I get a 404, not if there's a digest mismatch. I'm not sure what I'd recommend here, but IMO if `digest` doesn't match `data` for a given descriptor, I would probably just fail loudly and immediately, because I'm obviously dealing with a corrupt or untrustworthy artifact. Interested in others' opinions.

Reply:
> If a registry does want to allow skipping the upload for embedded content, then it SHOULD validate that the contents match

I think this is asking a fair amount from registry implementations, wouldn't they need to know how the client constructs the data from the layers in the case of multiple layers?

Reply:
> > If a registry does want to allow skipping the upload for embedded content, then it SHOULD validate that the contents match
> I think this is asking a fair amount from registry implementations,

To be clear, I think this is a bad idea:

> but I would be really reluctant to recommend this until we've let this change bake for a while (years).

It's something I could imagine happening one day, but it's not something I would recommend a registry actually do.

As a result of this PR, registries should literally change nothing (unless they currently reject unknown fields, which violates the extensibility expectations of this spec).

> wouldn't they need to know how the client constructs the data from the layers in the case of multiple layers

Not at all. This PR modifies the descriptor struct. The `data` field would only ever apply to one thing. It's also not specific to layers. Literally any piece of content that can be represented by a classical computer can be referenced by a descriptor, and the mechanism for constructing the `data` field is always the same: base64 encode it.